upgrade/upgrade-scripts/execute

```diff
@@ -323,6 +323,56 @@ zcat "/usr/share/doc/delphix-entire-$platform/packages.list.gz" |
 #
 apt_get autoremove --purge -y || die "autoremove after upgrade failed"
 
+#
+# Package configuration files are only automatically removed by the
+# package manager when the package that "owns" the file is "purged".
+# Thus, when upgrading a package to a new version that no longer
+# delivers a "conffile" that it used to deliver, that conffile will not
+# be automatically removed by the package manager.
+#
+# Generally, it is the responsibility of each specific package to
+# properly remove the conffile that is no longer used by the package.
+# Unfortuantely this means that if a given package does not do this
+# removal, the conffile will remain on the system indefinitely. Further,
+# that configuration file may actually result in a mis-configured
+# system.
+#
+# To help avoid mis-configuration when packages mistakenly omit the
+# removal of their conffiles, we have logic below to automatically
+# detect these "obsolete" conffiles, and remove them proactively.
+# Further, after we remove these obsolete files, we "--reinstall" the
+# package that previously owned the file, such that the package manager
+# will no longer report that file as obsolete.
+#
+dpkg-query -Wf '${Conffiles}\n' | awk '$3 == "obsolete" {print $1}' |
+	while read -r file; do
+		package=$(dpkg-query -S "$file" | awk -F: '{print $1}')
+		[[ -n "$package" ]] || die "failed to determine package name"
+
+		#
+		# We should only be operating on currently installed
+		# packages. So, to verify the package retrieved above,
+		# we double check that the package is installed.
+		#
+		dpkg-query -W "$package" &>/dev/null ||
+			die "package '$package' is not installed"
+
+		rm -f "$file" ||
+			die "failed to remove file '$file' of package '$package'"
+
+		#
+		# Since we run this each time a file is removed, if a
+		# package has multiple obsolete files, we'll end up
+		# reinstalling the package multiple times. Currently
+		# this isn't a problem, since we haven't hit a case
+		# where a package has multiple obsolete files; but
+		# perhaps this will change in the future, at which point
+		# we can optimize this better to handle that case.
+		#
+		apt_get install -y --reinstall "$package" ||
+			die "failed to reinstall package '$package'"
```

Comment:
Does `--reinstall` require the package deb to be available? While most of our package debs should be available during upgrade this would not be the case for packages pinned by `delphix-kernel`.

From my understanding, we "reinstall" the package just to remove those files from the "obsolete" list, as such should we just make this a warning instead of a failure?

Reply:
> Does --reinstall require the package deb to be available? While most of our package debs should be available during upgrade this would not be the case for packages pinned by delphix-kernel.

Hm, yea, I think this is a good point. I'll have to think about this case a little more...

> From my understanding, we "reinstall" the package just to remove those files from the "obsolete" list, as such should we just make this a warning instead of a failure?

Can you give me an example where we'd expect this to fail?

If there aren't examples where we expect this to fail, I'd rather not silently ignore the failure, without a good understanding about why it's safe/correct to ignore the failure. Just because we _can_ ignore the failure, doesn't mean  we _should_.

Reply:
I think the point about not being able to `--reinstall` packages that aren't in the upgrade image is valid, but I'm not sure if this will ever be a practical concern.. I think that a conffile will only become obsolete after a package is upgraded, and in order for the package to be upgraded it has to be contained in the upgrade image.. so, I think, any package that we find with an obsoleted confile _must_ be in the upgrade image, so we _must_ be able to `--reinstall` it.

By that logic, for the "old packages" pinned by "delphix-kernel", we will not have upgraded those packages, so I don't think we'll find any of their conffiles as being obsolete.. and thus, we won't attempt to `--reinstall` any of them.

What do you think? Am I mistaken, or overlooking anything?

Reply:
Following our conversation offline, I think it's fine to leave it as an error as chances that we would be trying to reinstall packages related to delphix-kernel are very small and a failure to reinstall could have potentially bad consequences, so it would be better that we catch it as an error in the upgrade container.

Reply:
> chances that we would be trying to reinstall packages related to delphix-kernel are very small

For posterity and just to be clear, I'm claiming the chances are _zero_ rather than _small_, and I think the distinction is important.